Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,600 @@
+        /// synchronization (see the `stateLock` property)
+        fileprivate var state: State
+
+        /// The lock used to access and modify the `state` property
+        ///
+        /// - Warning: This lock should always be acquired *after* `ConnectionPool`s `connectionProvidersLock` if used in combination with it.
+        fileprivate let stateLock = Lock()
+
+        /// The maximum number of concurrent connections to a given (host, scheme, port)
+        private let maximumConcurrentConnections: Int = 8
+
+        /// The pool this provider belongs to
+        private let parentPool: ConnectionPool
+
+        /// Creates a new `HTTP1ConnectionProvider`
+        ///
+        /// - parameters:
+        ///     - key: The `Key` (host, scheme, port) this provider is associated to
+        ///     - configuration: The client configuration used globally by all requests
+        ///     - initialConnection: The initial connection the pool initializes this provider with
+        ///     - parentPool: The pool this provider belongs to
+        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
+            self.eventLoop = eventLoop
+            self.configuration = configuration
+            self.key = key
+            self.parentPool = parentPool
+            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+        }
+
+        deinit {
+            assert(self.state.availableConnections.isEmpty, "Available connections should be empty before deinit")
+            assert(self.state.leased == 0, "All leased connections should have been returned before deinit")
+        }
+
+        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
+            self.preconditionIsOpened()
+            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
+            switch action {
+            case .leaseConnection(let connection):
+                return self.eventLoop.makeSucceededFuture(connection)
+            case .makeConnection(let eventLoop):
+                return self.makeConnection(on: eventLoop)
+            case .leaseFutureConnection(let futureConnection):
+                return futureConnection
+            }
+        }
+
+        func release(connection: Connection) {
+            self.preconditionIsOpened()
+            let action = self.parentPool.connectionProvidersLock.withLock {
+                self.stateLock.withLock { self.state.releaseAction(for: connection) }
+            }
+            switch action {
+            case .succeed(let promise):
+                promise.succeed(connection)
+
+            case .makeConnectionAndComplete(let eventLoop, let promise):
+                self.makeConnection(on: eventLoop).cascade(to: promise)
+
+            case .replaceConnection(let eventLoop):
```

Comment:
Code coverage indicates this is never called by tests, must add a test exercising this